internal/factsengine/gatherers/packageversion.go

```diff
@@ -120,13 +135,71 @@ func executeZypperVersionCmpCommand(
 func executeRpmVersionRetrieveCommand(
 	executor utils.CommandExecutor,
 	packageName string,
-) (string, *entities.FactGatheringError) {
-	rpmOutput, err := executor.Exec("/usr/bin/rpm", "-q", "--qf", "%{VERSION}", packageName)
+) ([]packageVersion, *entities.FactGatheringError) {
+	rpmOutputBytes, err := executor.Exec("/usr/bin/rpm", "-q", "--qf", packageVersionQueryFormat, packageName)
+
+	rpmOutput := string(rpmOutputBytes)
+
 	if err != nil {
-		gatheringError := PackageVersionRpmCommandError.Wrap(string(rpmOutput) + err.Error())
+		gatheringError := PackageVersionRpmCommandError.Wrap(rpmOutput + err.Error())
 		log.Error(gatheringError)
-		return "", gatheringError
+		return nil, gatheringError
+	}
+
+	installedVersions := []packageVersion{}
+
+	for _, detectedVersionLine := range strings.Split(rpmOutput, "\n---\n") {
+		if detectedVersionLine == "" {
+			continue
+		}
+
+		packageVersionInfo, err := envparse.Parse(strings.NewReader(detectedVersionLine))
+
+		if err != nil {
+			parsingError := fmt.Sprintf("Unable to parse rpm output: %s, output:%s", err.Error(), rpmOutput)
+			return nil, PackageVersionRpmCommandError.Wrap(parsingError)
+		}
+
+		packageVersionRequiredKeys := []string{installationTimeKey, packageVersionKey}
+
+		detectedPackageInstallationTime, installationTimeFound := packageVersionInfo[installationTimeKey]
+		detectedPackageVersion, packageVersionFound := packageVersionInfo[packageVersionKey]
+
+		if !installationTimeFound || !packageVersionFound {
+			missingPackageInfoError := fmt.Sprintf(
+				"Unable to find required information (keys: %s) in rpm output:%s",
+				packageVersionRequiredKeys,
+				rpmOutput,
+			)
+			return nil, PackageVersionRpmCommandError.Wrap(missingPackageInfoError)
+		}
+
+		timestamp, err := strconv.ParseInt(detectedPackageInstallationTime, 10, 64)
+		if err != nil {
+			invalidDateError := fmt.Sprintf("Unable to parse package installation timestamp to an integer: %s", err.Error())
+			return nil, PackageVersionRpmCommandError.Wrap(invalidDateError)
+		}
+
+		installedVersions = append(installedVersions, packageVersion{
+			Version:     detectedPackageVersion,
+			InstalledOn: time.Unix(timestamp, 0).UTC(),
+		})
+	}
+
+	sort.Slice(installedVersions, func(i, j int) bool {
```

Comment:
Maybe, you simplify this by adding `--last` to the rpm query.
`--last Orders the package listing by install time such that the latest packages are at the top.`

Reply:
Yes, that made sense to me as well, but unfortunately I discovered that the combination of `--last` with a `--qf` outputs something that would add unnecessary complexity to the output parsing.

```bash
$ rpm -q --qf "VERSION=%{VERSION}\nINSTALLTIME=%{INSTALLTIME}\n---\n" --last sbd
sbd-1.5.1+20221128.8ec8e01-150400.3.3.1.x86_64 Thu Feb  9 11:02:56 2023
sbd-1.5.1+20211116.6bb085f-150400.1.5.x86_64  Thu Feb  9 11:02:14 2023
VERSION=1.5.1+20221128.8ec8e01                
VERSION=1.5.1+20211116.6bb085f                
INSTALLTIME=1675940576                        
INSTALLTIME=1675940534                        
---                                           
---
```

The alternative would be not using a query format, losing its benefits
```bash
$ rpm -q --last sbd
sbd-1.5.1+20221128.8ec8e01-150400.3.3.1.x86_64 Thu Feb  9 11:02:56 2023
sbd-1.5.1+20211116.6bb085f-150400.1.5.x86_64  Thu Feb  9 11:02:14 2023
```

I went for a middle ground.